### ISSUE TITLE
No WC_Cart in non-frontend requests

### ISSUE THREAD

**Issue by barrykooij** [woothemes/woocommerce#9251]
WooCommerce only setups the WC_Cart class when the request is a frontend request, code:
```
if ( $this->is_request( 'frontend' ) ) {
$this->cart     = new WC_Cart(); // Cart class, stores the cart contents
$this->customer = new WC_Customer();  // Customer class, handles data such as customer location
}
```

The cart class contains methods specific to the current cart but also methods that are more general methods like `get_cart_url()` and `get_checkout_url()`. I've got a case where I want to include a modified cart ULR in a automatically send email but because this request is ran via wp-cron no `WC_Cart` object is created.

Also, because `WC_Cart` adds all sort of actions in it's `_construct()` I can't simply create my own `WC_Cart` object.

I think there are multiple ways of fixing this:

1. Separate 'helper'-like methods to own class like `WC_Cart_Helper`. This could be contain methods like     `get_cart_url()` and should always be available.
2. Move adding of hooks aways from constructor to an `init()` method, always initiate `WC_Cart` object in main WooCommerce class but only run `init()` method when it's a frontend request.

Looking forwards to see your thoughts on this.

**Comment by maxrice** [comment]
:+1: for option 2. We've run into issues with this in the past with payment gateway code that operates in the admin (refunds, capture charge, etc) trying to clear the cart and causing a fatal error.

**Comment by justinshreve** [comment]
:+1: for something like option 1 - I could see the URLs that are not session related being separate and usable. That totally makes sense. but :-1: for option 2. Option 2 doesn't make much sense because WC_Cart contains the current user/session's cart. What would that be in the admin?

In the mean time, for the cart URL in particular you can just do `wc_get_page_permalink( 'cart' )`.

**Comment by barrykooij** [comment]
I would personally prefer option 1 as well.

**Comment by claudiosmweb** [comment]
I think we will go for option number 3 xD
Will be like option 1, but create new functions to it, as have other functions like `wc_empty_cart()` and so on in https://github.com/woothemes/woocommerce/blob/master/includes/wc-cart-functions.php

**Comment by claudiosmweb** [comment]
Closed in favor of #9457
